src/core/pipe/pipe.middleware.js

```diff
@@ -0,0 +1,37 @@
+export default class Middleware {
+
+	constructor(stages) {
+		this.stages = stages || [];
+	}
+
+	run(ctx, data = []) {
+		return pipeline(
+			this.stages.map(s => data => new Promise(resolve => {
+				s(data, ctx, resolve);
+			})),
+			data);
+	}
+}
+
+function pipeline(stages, initialData) {
```

Comment:
can we reaname initialData to rows? so it'll show what type of data we are expecting on the start?)